be/src/olap/memtable.cpp

```diff
@@ -267,79 +218,137 @@ void MemTable::_aggregate_two_row_in_block(RowInBlock* new_row, RowInBlock* row_
-                      auto value = (*(this->_vec_row_comparator))(l, r);
-                      if (value == 0) {
-                          return l->_row_pos > r->_row_pos;
-                      } else {
-                          return value < 0;
-                      }
-                  });
-        std::vector<int> row_pos_vec;
-        DCHECK(in_block.rows() <= std::numeric_limits<int>::max());
-        row_pos_vec.reserve(in_block.rows());
-        for (int i = 0; i < _row_in_blocks.size(); i++) {
-            row_pos_vec.emplace_back(_row_in_blocks[i]->_row_pos);
-        }
-        _output_mutable_block.add_rows(&in_block, row_pos_vec.data(),
-                                       row_pos_vec.data() + in_block.rows());
-    } else {
-        size_t idx = 0;
-        for (it.SeekToFirst(); it.Valid(); it.Next()) {
-            auto& block_data = in_block.get_columns_with_type_and_name();
-            // move key columns
-            for (size_t i = 0; i < _schema->num_key_columns(); ++i) {
-                _output_mutable_block.get_column_by_position(i)->insert_from(
-                        *block_data[i].column.get(), it.key()->_row_pos);
+    vectorized::MutableBlock mutable_block =
+            vectorized::MutableBlock::build_mutable_block(&in_block);
+    _vec_row_comparator->set_block(&mutable_block);
+    std::vector<RowInBlock*> temp_row_in_blocks;
+    std::vector<RowInBlock*> agg_row_in_temp;
+    RowInBlock* prev_row;
+    temp_row_in_blocks.reserve(_last_sorted_pos);
+    agg_row_in_temp.reserve(_last_sorted_pos);
+    bool need_init_agg = true;
+    //only init agg if needed
+    for (int i = 0; i < _row_in_blocks.size(); i++) {
+        if (!temp_row_in_blocks.empty() &&
+            (*_vec_row_comparator)(prev_row, _row_in_blocks[i]) == 0) {
+            if (need_init_agg) {
+                prev_row->init_agg_places(
+                        _arena->aligned_alloc(_total_size_of_aggregate_states, 16),
+                        _offsets_of_aggregate_states.data());
+                for (auto cid = _schema->num_key_columns(); cid < _schema->num_columns(); cid++) {
+                    auto col_ptr = _input_mutable_block.mutable_columns()[cid].get();
+                    auto data = prev_row->agg_places(cid);
+                    _agg_functions[cid]->create(data);
+                    _agg_functions[cid]->add(
+                            data, const_cast<const doris::vectorized::IColumn**>(&col_ptr),
+                            prev_row->_row_pos, nullptr);
+                }
+                agg_row_in_temp.push_back(prev_row);
+                need_init_agg = false;
             }
+            _merged_rows++;
+            _aggregate_two_row_in_block(_row_in_blocks[i], prev_row);
+        } else {
+            prev_row = _row_in_blocks[i];
+            temp_row_in_blocks.push_back(prev_row);
+            need_init_agg = true;
+        }
+    }
```

Comment:
we should stat merged rows in one run and we can optimize move operations below.